File: packages/python/server/starfish_server/storage/k2v_adapter.py
"""KVAdapter backed by Garage K2V.

Mirror of ``packages/ts/server/src/storage/k2v-adapter.ts``. See that file's header for the
full consistency rationale. K2V has NO compare-and-set, NO atomic increment, and NO native
key expiry: concurrent writes become "siblings" merged by the reader via an
``X-Garage-Causality-Token``. This adapter therefore:

  - embeds an ``exp`` (ms) in every value and treats expired-on-read entries as absent
    (logically expired keys linger in K2V until overwritten or externally pruned);
  - on ``increment``, sums all live siblings and writes back the merged total superseding
    the read causality token. Concurrent increments may briefly **overcount** (stricter,
    fail-closed) — never undercount;
  - on ``record_if_absent``, does a best-effort read-then-write (no CAS → a narrow
    concurrent-duplicate replay window). The ``group`` fail-closed cap is **ignored**.

The HTTP/auth boundary is an injectable transport, so the protocol logic here is testable
with a mock and you can supply auth (AWS SigV4, a proxy, etc.) however you like.
"""

from __future__ import annotations

import json
import time
from dataclasses import dataclass
from typing import Callable, Protocol


@dataclass
class K2VReadResult:
    """Result of a K2V ReadItem: the live + tombstone sibling values and causality token."""

    values: list[str]
    causality: str | None


class K2VTransport(Protocol):
    """The HTTP/auth boundary for :func:`create_k2v_adapter`."""

    async def read(self, partition_key: str, sort_key: str) -> K2VReadResult:
        """ReadItem ``(partition_key, sort_key)`` → sibling values + causality token."""
        ...

    async def insert(
        self, partition_key: str, sort_key: str, value: str, causality: str | None
    ) -> None:
        """InsertItem, superseding ``causality`` when provided."""
        ...


def _parse_value(raw: str) -> dict | None:
    try:
        v = json.loads(raw)
        return v if isinstance(v, dict) and isinstance(v.get("exp"), (int, float)) else None
    except (ValueError, TypeError):
        return None


class _K2VAdapter:
    def __init__(
        self,
        transport: K2VTransport,
        partition_key: str = "starfish-kv",
        now: Callable[[], float] | None = None,
    ) -> None:
        self._t = transport
        self._pk = partition_key
        self._now = now or (lambda: time.time() * 1000)

    async def increment(self, key: str, ttl_ms: int) -> int:
        result = await self._t.read(self._pk, key)
        t = self._now()
        total = 0
        for raw in result.values:
            v = _parse_value(raw)
            if v is not None and v["exp"] > t:
                total += int(v.get("n", 0))
        nxt = total + 1
        await self._t.insert(
            self._pk, key, json.dumps({"exp": t + ttl_ms, "n": nxt}), result.causality
        )
        return nxt

    async def record_if_absent(
        self, key: str, ttl_ms: int, group: tuple[str, int] | None = None
    ) -> bool:
        result = await self._t.read(self._pk, key)
        t = self._now()
        live = any(
            (v := _parse_value(raw)) is not None and v["exp"] > t for raw in result.values
        )
        if live:
            return False
        await self._t.insert(self._pk, key, json.dumps({"exp": t + ttl_ms}), result.causality)
        return True


def create_k2v_adapter(
    transport: K2VTransport,
    *,
    partition_key: str = "starfish-kv",
    now: Callable[[], float] | None = None,
) -> "_K2VAdapter":
    """Build a KVAdapter over Garage K2V. See the module header for consistency caveats
    (overcount under contention; best-effort replay protection without CAS)."""
    return _K2VAdapter(transport, partition_key, now)


__all__ = ["K2VTransport", "K2VReadResult", "create_k2v_adapter"]

File: packages/ts/server/src/storage/k2v-adapter.ts
/**
 * `KVAdapter` backed by Garage K2V (https://garagehq.deuxfleurs.fr/documentation/reference-manual/k2v/).
 *
 * This shares rate-limit counters and nonce flags across server instances. K2V is a
 * CRDT-style store with NO compare-and-set and NO atomic increment: concurrent writes to
 * one key become "siblings" that the reader merges, with optimistic concurrency via an
 * `X-Garage-Causality-Token`. K2V also has no native key expiry. This adapter therefore:
 *
 *  - embeds an `exp` (ms) in every value and treats expired-on-read entries as absent
 *    (logically expired keys linger in K2V until overwritten or externally pruned — there
 *    is no TTL; size/scrub your bucket accordingly);
 *  - on `increment`, reads all live siblings, **sums** their counts, and writes back the
 *    merged total superseding the read causality token. Under concurrent increments the
 *    siblings can briefly **overcount** (a stricter, fail-closed limit) — never undercount;
 *  - on `recordIfAbsent`, does a best-effort read-then-write. Without CAS, two *concurrent*
 *    requests with the *same* key can both be accepted (a narrow replay window). The `group`
 *    fail-closed cap is **ignored** (K2V can't cheaply count a group). Use a CAS-capable
 *    store if you need exact distributed replay protection.
 *
 * The HTTP/auth boundary is an injectable {@link K2VTransport}, so the protocol logic here
 * is testable with a mock and you can supply auth (AWS SigV4, a proxy, etc.) however you like.
 */
import type { KVAdapter } from "./kv-adapter.js"

/** Result of a K2V ReadItem: the live + tombstone sibling values and the causality token. */
export interface K2VReadResult {
  /** Sibling values (UTF-8). Tombstones are omitted. Empty when the key is absent. */
  values: string[]
  /** Opaque causality token to pass to a superseding insert/delete; null when absent. */
  causality: string | null
}

/** The HTTP/auth boundary for {@link createK2VAdapter}. Implement against Garage's K2V API. */
export interface K2VTransport {
  /** ReadItem `(partitionKey, sortKey)` → sibling values + causality token. */
  read(partitionKey: string, sortKey: string): Promise<K2VReadResult>
  /** InsertItem `(partitionKey, sortKey, value)`, superseding `causality` when provided. */
  insert(partitionKey: string, sortKey: string, value: string, causality: string | null): Promise<void>
}

/** Options for {@link createK2VAdapter}. */
export interface K2VAdapterOptions {
  /** The K2V HTTP/auth transport. */
  transport: K2VTransport
  /** K2V partition key under which all keys are stored (sort key = the app key).
   *  Default `"starfish-kv"`. */
  partitionKey?: string
  /** Clock (ms); defaults to `Date.now`. */
  now?: () => number
}

interface StoredValue {
  exp: number
  n?: number
}

function parseValue(raw: string): StoredValue | null {
  try {
    const v = JSON.parse(raw) as StoredValue
    return typeof v?.exp === "number" ? v : null
  } catch {
    return null
  }
}

/**
 * Build a {@link KVAdapter} over Garage K2V. See the file header for the consistency
 * caveats (overcount under contention; best-effort replay protection without CAS).
 */
export function createK2VAdapter(opts: K2VAdapterOptions): KVAdapter {
  const transport = opts.transport
  const pk = opts.partitionKey ?? "starfish-kv"
  const now = opts.now ?? Date.now

  return {
    async increment(key, ttlMs) {
      const { values, causality } = await transport.read(pk, key)
      const t = now()
      let sum = 0
      for (const raw of values) {
        const v = parseValue(raw)
        if (v && v.exp > t) sum += v.n ?? 0
      }
      const next = sum + 1
      await transport.insert(pk, key, JSON.stringify({ exp: t + ttlMs, n: next }), causality)
      return next
    },

    async recordIfAbsent(key, ttlMs, _group) {
      const { values, causality } = await transport.read(pk, key)
      const t = now()
      const live = values.some((raw) => {
        const v = parseValue(raw)
        return v != null && v.exp > t
      })
      if (live) return false
      await transport.insert(pk, key, JSON.stringify({ exp: t + ttlMs }), causality)
      return true
    },
  }
}

// --- Default fetch-based transport ---

/** Hook to attach auth (e.g. AWS SigV4) to an outgoing K2V request before it is sent. */
export type K2VRequestSigner = (req: {
  method: string
  url: string
  headers: Record<string, string>
  body?: string
}) => Record<string, string> | Promise<Record<string, string>>

/** Options for {@link createFetchK2VTransport}. */
export interface FetchK2VTransportOptions {
  /** Base URL of the Garage K2V endpoint, e.g. `"https://garage.example.com"`. */
  endpoint: string
  /** Bucket name. */
  bucket: string
  /** `fetch` implementation. Defaults to the global `fetch`. */
  fetch?: typeof fetch
  /** Optional hook to add auth headers (SigV4, bearer, etc.). When omitted, requests are
   *  sent unsigned — only appropriate for a trusted network or an auth-injecting proxy. */
  signRequest?: K2VRequestSigner
}

/**
 * A {@link K2VTransport} over HTTP using `fetch`, speaking the Garage K2V API: ReadItem
 * with `Accept: application/json` (returns a base64 sibling array + `X-Garage-Causality-Token`),
 * and InsertItem via `PUT ?sort_key=` carrying the causality token to supersede.
 */
export function createFetchK2VTransport(opts: FetchK2VTransportOptions): K2VTransport {
  const doFetch = opts.fetch ?? fetch
  const base = opts.endpoint.replace(/\/$/, "")
  const enc = (s: string): string => encodeURIComponent(s)

  async function signed(method: string, url: string, body?: string): Promise<Record<string, string>> {
    const headers: Record<string, string> = {}
    if (body !== undefined) headers["Content-Type"] = "application/octet-stream"
    if (opts.signRequest) return { ...headers, ...(await opts.signRequest({ method, url, headers, body })) }
    return headers
  }

  return {
    async read(partitionKey, sortKey) {
      const url = `${base}/${enc(opts.bucket)}/${enc(partitionKey)}?sort_key=${enc(sortKey)}`
      const headers = await signed("GET", url)
      headers["Accept"] = "application/json"
      const res = await doFetch(url, { method: "GET", headers })
      if (res.status === 404) return { values: [], causality: null }
      if (!res.ok) throw new Error(`K2V read failed: ${res.status}`)
      const causality = res.headers.get("x-garage-causality-token")
      // ReadItem with Accept: application/json returns a JSON array of base64 strings;
      // a `null` entry is a tombstone (concurrent delete) and is skipped.
      const arr = (await res.json()) as (string | null)[]
      const values = arr
        .filter((v): v is string => v != null)
        .map((b64) => Buffer.from(b64, "base64").toString("utf-8"))
      return { values, causality }
    },

    async insert(partitionKey, sortKey, value, causality) {
      const url = `${base}/${enc(opts.bucket)}/${enc(partitionKey)}?sort_key=${enc(sortKey)}`
      const headers = await signed("PUT", url, value)
      if (causality) headers["X-Garage-Causality-Token"] = causality
      const res = await doFetch(url, { method: "PUT", headers, body: value })
      if (!res.ok) throw new Error(`K2V insert failed: ${res.status}`)
    },
  }
}
